helper/schema/schema.go

```diff
@@ -1431,6 +1480,79 @@ func (m schemaMap) validateConflictingAttributes(
 	return nil
 }
 
+func removeDuplicates(elements []string) []string {
+	encountered := make(map[string]struct{}, 0)
+	result := []string{}
+
+	for v := range elements {
+		if _, ok := encountered[elements[v]]; !ok {
+			encountered[elements[v]] = struct{}{}
+			result = append(result, elements[v])
+		}
+	}
+
+	return result
+}
+
+func validateExactlyOneAttributes(
+	k string,
+	schema *Schema,
+	c *terraform.ResourceConfig) error {
+
+	if len(schema.ExactlyOneOf) == 0 {
+		return nil
+	}
+
+	allKeys := removeDuplicates(append(schema.ExactlyOneOf, k))
+	sort.Strings(allKeys)
+	specified := make([]string, 0)
+	for _, exactlyOneOfKey := range allKeys {
+		if raw, ok := c.Get(exactlyOneOfKey); ok {
+			if raw == hcl2shim.UnknownVariableValue {
+				// An unknown value might become unset (null) once known, so
+				// we must defer validation until it's known.
+				continue
+			}
+			specified = append(specified, exactlyOneOfKey)
+		}
+	}
+
+	if len(specified) == 0 {
+		return fmt.Errorf("%q: one of `%s` must be specified", k, strings.Join(allKeys, ","))
+	}
+
+	if len(specified) > 1 {
+		return fmt.Errorf("%q: only one of `%s` can be specified, but `%s` were specified.", k, strings.Join(allKeys, ","), strings.Join(specified, ","))
+	}
+
+	return nil
+}
+
+func validateAtLeastOneAttributes(
+	k string,
+	schema *Schema,
+	c *terraform.ResourceConfig) error {
+
+	if len(schema.AtLeastOneOf) == 0 {
+		return nil
+	}
+
+	allKeys := removeDuplicates(append(schema.AtLeastOneOf, k))
```

Comment:
my brain is tingling about this LOC, but if the tests pass I'm sure it all makes sense.

Reply:
This was added to ensure that users coming from the `ConflictsWith` pattern where you specify every key but the attribute you're in are covered. For example:
```
"foo": {
  ConflictsWith: ["bar"]
}
"bar": {
  ConflictsWith:["foo"]
}
```
We wanted users to go a different route where you specify the whole list:
```
"foo": {
  AtLeastOneOf: ["foo", "bar"]
}
"bar": {
  AtLeastOneOf:["foo", "bar"]
}
```
but we don't have anyway to enforce that they use this pattern so we added this removal to ensure that aren't looping over any duplicates that cause our validation to go sideways